pycozmo/conn.py

```diff
@@ -158,19 +168,35 @@ def run(self) -> None:
             except Empty:
                 continue
 
+            raw_frames = []
+
             # Construct frame
             if isinstance(pkt, protocol_encoder.Ping):
-                frame = Frame(protocol_declaration.FrameType.PING, 0, 0, self.last_ack, [pkt])
+                raw_frames.append(
+                    Frame(protocol_declaration.FrameType.PING, 0, 0, self.last_ack, [pkt])
+                    .to_bytes())
             else:
-                seq = self.window.put(pkt)
-                if pkt.type == PacketType.COMMAND:
-                    frame = Frame(protocol_declaration.FrameType.ENGINE_ACT, seq, seq, self.last_ack, [pkt])
-                else:
-                    frame = Frame(protocol_declaration.FrameType.ENGINE, seq, seq, self.last_ack, [pkt])
-            raw_frame = frame.to_bytes()
+                first_seq, pkts = self.window.put(pkt)
+                framelen = 0
+                seq = first_seq
+                pkts = []
+                for p in pkts:
+                    framelen += len(p) + 1
+                    if framelen < 1000:
+                        pkts.append(pkt)
+                        seq += 1
+                    else:
+                        if len(pkts) == 1:
+                            frame = Frame(protocol_declaration.FrameType.ENGINE_ACT, first_seq, seq, self.last_ack, pkts)
+                        else:
+                            frame = Frame(protocol_declaration.FrameType.ENGINE, first_seq, seq, self.last_ack, pkts)
+                        raw_frames.append(frame.to_bytes())
+                        pkts = []
+                        framelen = 0
 
             try:
-                self.sock.sendto(raw_frame, self.receiver_address)
+                for raw_frame in raw_frames:
```

Comment:
I love the fact that you are working on sending multiple packets with one `Frame`.

The current implementation though, I think will resend all unacknowledged packets on every new packets.

The goal instead should be to look into the `SendThread` queue, pack whatever is in it into frames and send these.

Reply:
Yep, I agree, that would be better. I can also include a special case for the ping so it's sent instantly instead of wait for the time to send the packets. We could have a while loop that empties the queue and sends the available packets on a timeout, but if a ping is received then we send it without waiting or affecting the timeout.